### ISSUE TITLE
Cannot spread object literal into object with index signature

### ISSUE THREAD

**Issue by ajafff** [Microsoft/TypeScript#16373]
**TypeScript Version:**  nightly (2.4.0-dev.20170608)

**Code**

```ts
declare let compilerOptions: ts.CompilerOptions;
const options: ts.CompilerOptions = {
    ...compilerOptions,
    noEmit: true,
    noUnusedLocals: true,
    ...{noUnusedParameters: true}, // was originally ...(condition ? {noUnusedParameters: true} : undefined)
};
```

**Expected behavior:**
No error as with typescript@2.3.4

**Actual behavior:**
```
error TS2322: Type '{ noEmit: true; noUnusedLocals: true; allowJs?: boolean | undefined; allowSyntheticDefaultImports...' is not assignable to type 'CompilerOptions'.
  Type '{ noEmit: true; noUnusedLocals: true; allowJs?: boolean | undefined; allowSyntheticDefaultImports...' is not assignable to type 'CompilerOptions'.
    Index signature is missing in type '{ noEmit: true; noUnusedLocals: true; allowJs?: boolean | undefined; allowSyntheticDefaultImports...'.
```

To make this work, I need to assign the object literal to a variable with a type annotation containing an index signature.

**Comment by sandersn** [comment]
I can't get this to repro inside our own code. Can you try this again on `typescript@next`? I'll see if I can get a standalone repro that doesn't depend on ts.CompilerOptions.

**Comment by sandersn** [comment]
Nope, standalone code (not depending on ts.CompilerOptions) doesn't repro either.

**Comment by ajafff** [comment]
Seems to be "fixed" by #16525
Will open a new issue if this bug appears again after the handling of `undefined` in spread types changes.
